packages/contracts/contracts/core/DAO.sol

```diff
@@ -215,47 +248,156 @@ contract DAO is
+                    if (_amount != 0) {
+                        revert NonZeroAmount({amount: _amount});
+                    }
+                    IERC721Upgradeable(_token).safeTransferFrom({
+                        from: msg.sender,
+                        to: address(this),
+                        tokenId: _tokenId
+                        //, data: _reference // TODO Should we pass the data or save gas because we emit it in the event anyway.
+                    });
+                }
+                // -- ERC-1155 Token -- //
+                else if (_token.supportsInterface(type(IERC1155Upgradeable).interfaceId)) {
+                    if (_amount == 0) {
+                        revert ZeroAmount();
+                    }
+                    IERC1155Upgradeable(_token).safeTransferFrom({
+                        from: msg.sender,
+                        to: address(this),
+                        id: _tokenId,
+                        amount: _amount,
+                        data: bytes(_reference)
+                    });
+                } else {
+                    revert UnsupportedTokenStandard({token: _token});
+                }
+            }
+            // -- ERC-20 Token -- //
+            else {
+                if (_tokenId != 0) {
+                    revert NonZeroTokenId({tokenId: _tokenId});
+                }
+                if (_amount == 0) {
+                    revert ZeroAmount();
+                }
+                IERC20Upgradeable(_token).safeTransferFrom(msg.sender, address(this), _amount);
+            }
         }
 
-        emit Deposited(msg.sender, _token, _amount, _reference);
+        emit Deposited({
+            sender: msg.sender,
+            token: _token,
+            amount: _amount,
+            tokenId: _tokenId,
+            _reference: _reference
+        });
     }
 
     /// @inheritdoc IDAO
     function withdraw(
         address _token,
         address _to,
+        uint256 _tokenId,
         uint256 _amount,
         string memory _reference
-    ) external override auth(address(this), WITHDRAW_PERMISSION_ID) {
-        if (_amount == 0) revert ZeroAmount();
-
+    ) external auth(address(this), WITHDRAW_PERMISSION_ID) {
+        // -- Native Token -- //
         if (_token == address(0)) {
+            if (_amount == 0) {
+                revert ZeroAmount();
+            }
+            if (_tokenId != 0) {
+                revert NonZeroTokenId({tokenId: _tokenId});
+            }
             (bool ok, ) = _to.call{value: _amount}("");
-            if (!ok) revert NativeTokenWithdrawFailed();
-        } else {
-            IERC20Upgradeable(_token).safeTransfer(_to, _amount);
+            if (!ok) {
+                revert NativeTokenWithdrawFailed();
+            }
+        }
+        // -- Non-Native Tokens -- //
+        else {
+            // -- ERC-165 Supporting Tokens -- //
+            if (_token.supportsERC165()) {
+                // -- ERC-721 Token -- //
```

Comment:
These comments make more uglier ! the code directly suggests what it does. If it checks `supportsInterface(type(IERC721Upgradeable).interfaceId)`, then reader knows it's about ERC721.